packages/ui-extensions/src/surfaces/checkout/api/shared.ts

```diff
@@ -148,3 +148,34 @@ export interface ShippingAddress extends MailingAddress {
    */
   oneTimeUse?: boolean;
 }
+
+type MaybePromise<T> = T extends Promise<any> ? T : T | Promise<T>;
+type Subscriber<T> = (value: T) => unknown;
+type RemoteSubscribeResult<T> = [() => void, T];
+
+/**
+ * A remote-subscribable object managed on the main thread.
+ *
+ * Example: Checkout uses this to manage the state of an object and
+ * communicate state changes to extensions running in a sandboxed web worker.
+ */
+export interface RemoteSubscribable<T> {
+  readonly initial: T;
+  subscribe(subscriber: Subscriber<T>): MaybePromise<RemoteSubscribeResult<T>>;
+}
+
+/**
+ * A remote-subscribable object exposed to a sandboxed web worker.
+ *
+ * Example: extensions use this to get updates about an object
+ * managed by Checkout on the main thread.
+ *
+ * This interface is compatible with Preact's ReadonlySignal:
+ * https://github.com/preactjs/signals/blob/a023a132a81bd4ba4a0bebb8cbbeffbd8c8bbafc/packages/core/src/index.ts#L700-L709
+ */
+export interface StatefulRemoteSubscribable<T> {
+  readonly value: T;
+  readonly current: T;
+  subscribe(subscriber: Subscriber<T>): () => void;
+  destroy(): Promise<void>;
```

Comment:
Checkout is the only surface that needs a backwards compatible subscribable. Let's keep it in Checkout but expose the shared type in the new aligned-on shape ([private alignment spec](https://github.com/Shopify/ui-api-design/blob/main/ui-api-design/shared/subscribable.ts)). I agree with the earlier comments that we should pick a new name 🚲🚪 I suggest a name that communicates how it's closely related to a signal:

```ts
// In src/shared.ts
export interface ReadonlySignalLike<T> {
  readonly value: T;
  subscribe(subscriber: Subscriber<T>): () => void;
}

// In src/surfaces/checkout/api/shared.ts
export interface StatefulRemoteSubscribable<T> extends ReadonlySignalLike<T> {
  readonly current: T;
  destroy(): Promise<void>;
}
```